File: 202-happy-number/202-happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        set = {n}
        while n!= 1:
            n = int(sum(int(d)**2 for d in str(n)))
            if n in set:
                return False
            set.add(n)
        
        return True

File: 202-happy-number/202-happy-number.cpp
class Solution {
public:
    bool isHappy(int n) {
        int slow = n, fast = n;
        slow = square_sum(slow);
        fast = square_sum(square_sum(fast));
        while(slow != fast){
            slow = square_sum(slow);
            fast = square_sum(square_sum(fast));
        }
        return slow == 1;
    }
    private:
        int square_sum(int n){
            int sum = 0;
            while(n>0){
                int d = n%10;
                sum += d*d;
                n/=10;
            }
            return sum;
        }
};